Multi-GPU Training.md

```diff
@@ -0,0 +1,89 @@
+# Multi-GPU Training
+
+There are 2 ways you can leverage multiple GPUs to train DeepLIIF, **Data Parallel (DP)** or **Distributed Data Parallel (DDP)**. Both cases are a kind of **data parallelism** supported by PyTorch.
+
+The key difference is that DP is **single process multi-threading** while DDP can have **multiple processes**.
+
+**TL;DR**
+
+Use DP if you
+- are used to the way to train DeepLIIF on multiple GPUs since its first release, OR
+- do **not** have multiple GPU machines to utilize, OR
+- are fine with the training being a bit slower
+
+Use DDP if you
+- are willing to try a slightly different way to launch the training than before, OR
+- do have multiple GPU machines for cross-node distributino, OR
+- want to get as fast training as possible
+
+
+## Data Parallel (DP)
+DP is single-process. It means that **all the GPUs you want to use must be on the same machine** so that they can be included in the same process - you cannot distribute the training across multiple GPU machines, unless you write your own code to handle inter-node (node = machine) communication.
+
+To split and manage the workload for multiple GPUs within the same process, DP uses multi-threading. 
+
+It is worth noting that multi-threading in this case can lead to significant performance overhead, and slow down your training. See a short discussion in [PyTorch's CUDA Best Practices](https://pytorch.org/docs/stable/notes/cuda.html#use-nn-parallel-distributeddataparallel-instead-of-multiprocessing-or-nn-dataparallel).
+
+### Train with DP
+Example with 2 GPUs (of course on 1 machine):
+```
+python cli.py train --dataroot <data_dir> --batch-size 6 --gpu-ids 0 --gpu-ids 1
+```
+Note that
+1. `batch-size` is defined per process. Since DP is a single-process method, the `batch-size` you set is the **effective** batch size.
+
+## Distributed Data Parallel (DDP)
+DDP usually spawns multiple processes. 
+
+**DeepLIIF's code follows the PyTorch recommendation to spawn 1 process per GPU** ([doc](https://github.com/pytorch/examples/blob/master/distributed/ddp/README.md#application-process-topologies)). If you want to assign multiple GPUs to each process, you will need to make modifications to DeepLIIF's code (see [doc](https://pytorch.org/tutorials/intermediate/ddp_tutorial.html#combine-ddp-with-model-parallelism)).
+
+
+### Train with DDP
+#### 1. Local Machine
+To launch training using DDP on a local machine, simply add `torchrun --nproc_per_node <gpu_num>` before the training command. Example with 2 GPUs (on 1 machine):
```

Comment:
We should probably mention that `torchrun` is available is torch >= 1.10.0

Reply:
added a note for this and the alternative command if using older versions